### PR TITLE
Add mypy type checking

### PR DESCRIPTION
**Pull request by dstansby**
Since we're already using `black` in this repository, I thought it was worth experimenting with `mypy` here too. This PR adds basic mypy pre-commit config, and fixes all the issues it had.

### CODE REVIEW

File: sunkit_image/time_lag.py [hunk 1]
```diff
@@ -236,7 +238,7 @@ def time_lag(signal_a, signal_b, time: u.s, lag_bounds: (u.s, None) = None, **kw
 
 
 @u.quantity_input
-def max_cross_correlation(signal_a, signal_b, time: u.s, lag_bounds: (u.s, None) = None):
+def max_cross_correlation(signal_a, signal_b, time: u.s, lag_bounds: Optional[u.Quantity[u.s]] = None):
```
[the remark this answers is not in the document]
Comment (dstansby) [reply sunkit_image/time_lag.py:241]: Yep, see the third example at https://docs.astropy.org/en/stable/api/astropy.units.quantity_input.html#astropy.units.quantity_input

**Comment by nabobalis** [comment]
Oldest failure is:

```
../../.tox/py38-oldestdeps/lib/python3.8/site-packages/sunkit_image/tests/test_time_lag.py:7: in <module>
    from sunkit_image.time_lag import cross_correlation, get_lags, max_cross_correlation, time_lag
../../.tox/py38-oldestdeps/lib/python3.8/site-packages/sunkit_image/time_lag.py:167: in <module>
    def time_lag(signal_a, signal_b, time: u.s, lag_bounds: Optional[u.Quantity[u.s]] = None, **kwargs):
../../.tox/py38-oldestdeps/lib/python3.8/site-packages/astropy/units/quantity.py:422: in __class_getitem__
    warnings.warn(
E   UserWarning: Quantity annotations are valid static type annotations only if Python is v3.9+ or `typing_extensions` is installed.
```

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -37,5 +37,11 @@ repos:
     rev: v2.2.4
     hooks:
       - id: codespell
+
```
Comment (nabobalis) [inline .pre-commit-config.yaml:40]:
```suggestion
```

File: sunkit_image/tests/test_trace.py [hunk 3]
```diff
@@ -236,16 +236,6 @@ def test_erase_loop_in_image(image, test_map):
     assert np.allclose(expect, result)
 
 
-@pytest.fixture
-def test_image():
```
Comment (nabobalis) [inline sunkit_image/tests/test_trace.py:240]: Was this never used?

**Comment by nabobalis** [comment]
I think it's worth a changelog entry.

**Review by nabobalis — APPROVED** [review]

**Comment by nabobalis** [comment]
38-oldest is failing:
```
../../.tox/py38-oldestdeps/lib/python3.8/site-packages/sunkit_image/tests/test_time_lag.py:7: in <module>
    from sunkit_image.time_lag import cross_correlation, get_lags, max_cross_correlation, time_lag
../../.tox/py38-oldestdeps/lib/python3.8/site-packages/sunkit_image/time_lag.py:167: in <module>
    def time_lag(signal_a, signal_b, time: u.s, lag_bounds: Optional[u.Quantity[u.s]] = None, **kwargs):
../../.tox/py38-oldestdeps/lib/python3.8/site-packages/astropy/units/quantity.py:422: in __class_getitem__
    warnings.warn(
E   UserWarning: Quantity annotations are valid static type annotations only if Python is v3.9+ or `typing_extensions` is installed.
```
But I think it might be time to drop 38.

**Comment by dstansby** [comment]
Boo, I'll leave this until py38 is dropped then :thu

**Comment by nabobalis** [comment]
This PR was replaced by #156, I took the commits and merged in main.

**Pull request closed without merging**